File: demo/neural.py
from __future__ import division
from neural_data import *
from math import exp, e

# sigmoid function, used by the hidden neurons
# (runs over individual values)
def sig(n):
    return (2 / (1 + (e)**(-2 * n))) - 1

# softmax function, used by the output neurons
# (runs over an array)
# (also doesn't work properly right now (right results but scaled wrongly))
def softmax(data):
    exponentials = [exp(value) for value in data]
    denominator = sum(exponentials)
    return [numerator/denominator for numerator in exponentials]

# initial processing of inputs at the start of the ANN (pre-neurons)
#
# basically just scales inputs using offset and gain
def preprocess(data):
    data = [((data[input] - offset[input]) * gain[input] + ymin) for input in inputs]
    return data

# processing by the hidden layer of neurons
#
# takes the inputs, calculates a weighted sum over them, adds the neuron's bias,
# then pushes the result through the sigmoid function to get the output
#
# data: inputs to the neuron (list)
# neuron: index of the neuron
def hidden_layer(data, neuron):
    weighted_sum = sum([weights_hidden[neuron][input] * data[input] for input in inputs])
    return sig(bias_hidden[neuron] + weighted_sum)

# processing by the output layer of neurons
#
# takes inputs from hidden neurons, weights and biases them
# notably doesn't run softmax on them (softmax needs the
# data from all of the neurons, so is run later)
#
# data: inputs to the neuron (list)
# neuron: index of the neuron
def output_layer(data, neuron):
    weighted_sum = sum([weights_output[neuron][input] * data[input]
                        for input in hidden_neurons])
    return bias_output[neuron] + weighted_sum

# runs the full neural network over a dataset
# returns the list of outputs from the output neurons
def neural(data):
    data = preprocess(data)
    data = [hidden_layer(data, n) for n in hidden_neurons]
    data = [output_layer(data, n) for n in output_neurons]
    data = softmax(data)
    return data

# returns the name of the gesture calculated to be 'most likely' by the ANN
def best(data):
    data = neural(data)
    m = max(data)
    #return gestures[data.index(m)]
    return data

# names of gestures, indexed by the output neurons indices
gestures = [
    "idle", "push", "pull", "dragleft", "dragright",
    "hit", "lift", "screwin", "screwout"
]

File: API/Neural.cpp
#include <math.h>

#include "Neural.hpp"
#include "NeuralData.hpp"

/* Competitive Soft Transfer Function
 *
 * 'squeezes' the output of the neural network into unit vector (sums to one)
 */
static void softmax(int num, double* values) {
  double sum = 0.0;
  for (int i = 0; i < num; i++) {
    sum += exp(values[i]);
  }
  if (sum == 0.0) {
    sum = 1.0;
  }
  for (int i = 0; i < num; i++) {
    values[i] = exp(values[i]) / sum;
  }
}

/* Sigmoid Symmetric Transfer Function
 * 
 * specifically, a logistic function onto (-1, 1)
 */
static double sigmoid(double n) {
   return (2 / (1 + exp((-2) * n))) - 1;
}

/* calculates the vector dot product of two arrays
 *  
 *  num: length of the arrays
 *  v1: first vector
 *  v2: second vector
 */
static double dot_product(int num, const double* v1, const double* v2) {
	double r = 0.0;
	for (int i = 0; i < num; i++) {
		r += v1[i] * v2[i];
	}
	return r;
}

// gets weightings from NeuralData.cpp 
NeuralNetwork::NeuralNetwork(void) {
  gain = NeuralData::gain;
  offset = NeuralData::offset;
  yMin = NeuralData::yMin;
  weights_hidden = NeuralData::weights_hidden;
  bias_hidden = NeuralData::bias_hidden;
  weights_output = NeuralData::weights_output;
  bias_output = NeuralData::bias_output;
}

/* preprocessing for the ANN
 * scales each of the inputs to the network, then adds an offset
 */
void NeuralNetwork::preprocess(const double* inputs) const {
  for (int input = 0; input < NeuralData::input_num; input++) {
    preprocessed[input] = gain[input] * (inputs[input] - offset[input]) + yMin;
  }
}

/* hidden layer for the network
 * takes the preprocessed inputs, weights them and puts them through the sigmoid function
 */
void NeuralNetwork::hidden_layer(void) const {
  for (int neuron = 0; neuron < NeuralData::hidden_num; neuron++) {
    double sum = dot_product(NeuralData::input_num, weights_hidden[neuron], preprocessed);
    hidden_outputs[neuron] = sigmoid(sum + bias_hidden[neuron]);
  }
}

/* output layer for the network
 * weights the output from the hidden layer
 * doesn't have a non-linear component
 * (scaling the output from here is done by the calling code, usually using softmax)
 */
void NeuralNetwork::output_layer(double* outputs) const {
  for (int neuron = 0; neuron < NeuralData::output_num; neuron++) {
    double sum = dot_product(NeuralData::hidden_num, weights_output[neuron], hidden_outputs);
    outputs[neuron] = sum + bias_output[neuron];
  }
}

/* processes the given data through the neural network layers
 * 
 * inputs: input data, should be double[60]
 * outputs: output from the network, should be double[9]
 */
void NeuralNetwork::process(const double* inputs, double* outputs) const {
  preprocess(inputs);
  hidden_layer();
  output_layer(outputs);
  softmax(NeuralData::output_num, outputs);
}
